### PR TITLE
Add terraform drift detection

### LINKED ISSUE #2542: Automated configuration drift detection for terraform
Epic description
---

### PR DESCRIPTION
**Pull request by funera1**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2542

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by funera1** [comment]
There are two points of concern.
First, it does not use Cache, which would likely result in poor performance. However, terraform's drift detection is different from other kubernetes and cloud runs in the way drift detection is done, so implementing cache is also difficult.
Second, the other two output diff details in makeSyncState, but terraform does not. This is because PlanResult did not have a method called Render. However, it may be possible to prepare the details if we can get the output format of Render and the raw output of terraform plan

File: pkg/app/piped/driftdetector/terraform/detector.go [hunk 1]
```diff
@@ -13,3 +13,318 @@
...
+		m    = make(map[string][]*model.Application)
+	)
+	for _, app := range apps {
+		repoID := app.GitPath.Repo.Id
+		m[repoID] = append(m[repoID], app)
+	}
+	return m
+}
+
+func (d *detector) loadApplicationConfiguration(repoPath string, app *model.Application) (*config.Config, error) {
```
Comment (knanao) [inline pkg/app/piped/driftdetector/terraform/detector.go:316]:
```suggestion
func loadApplicationConfiguration(repoPath string, app *model.Application) (*config.Config, error) {
```

File: pkg/app/piped/driftdetector/terraform/detector.go [hunk 2]
```diff
@@ -13,3 +13,318 @@
...
+		appDir  = filepath.Join(repoDir, app.GitPath.Path)
+	)
+
+	// Load config
+	cpCfg := d.provider.TerraformConfig
+	cfg, err := d.loadApplicationConfiguration(repoDir, app)
+	appCfg := cfg.TerraformApplicationSpec
+	if err != nil {
+		return fmt.Errorf("failed to load application configuration: %w", err)
+	}
```
Comment (knanao) [inline pkg/app/piped/driftdetector/terraform/detector.go:186]:
```suggestion
	if err != nil {
		return fmt.Errorf("failed to load application configuration: %w", err)
	}
	appCfg := cfg.TerraformApplicationSpec
```

**Review by knanao — COMMENTED** [review]
Nice work.
I left some nits comments, so please check it out.
And let's call this feature to check whether working correctly.
https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/driftdetector/detector.go#L97

**Comment by funera1** [comment]
Thank you for your nice review

**Review by knanao — APPROVED** [review]

**Pull request merged**